base/strings/faces.jl

```diff
@@ -0,0 +1,779 @@
+
+Execute `f` with [`FACES`](@ref) temporarily modified by zero or more `:name =>
+val` arguments `kv`. `withfaces` is generally used via the `withfaces(kv...) do
+... end` syntax. A value of `nothing` can be used to temporarily unset an
+environment variable (if it is set). When `withfaces` returns, the original
+[`FACES`](@ref) has been restored.
+
+    !!! warning
+    Changing faces is not thread-safe.
+"""
+function withfaces(f, keyvals::Pair{Symbol, Union{Face, Symbol, Nothing}}...)
+    old = Dict{Symbol, Union{Face, Nothing}}()
+    for (name, face) in keyvals
+        old[name] = get(FACES, face, nothing)
+        if face isa Face
+            FACES[name] = face
+        elseif face isa Symbol
+            FACES[name] = get(FACES, face, Face())
+        elseif haskey(FACES, name)
+            delete!(FACES, name)
+        end
+    end
+    try f()
+    finally
+        for (name, face) in old
+            if isnothing(face)
+                delete!(FACES, name)
+            else
+                FACES[name] = face
+            end
+        end
+    end
+end
+
+"""
+    withfaces(f, altfaces::Dict{Symbol, Face})
+
+Execute `f` with [`FACES`](@ref) temporarily swapped out with `ALTFACES`
+When `withfaces` returns, the original [`FACES`](@ref) has been restored.
+
+    !!! warning
+    Changing faces is not thread-safe.
+"""
+function withfaces(f, altfaces::Dict{Symbol, Face})
+    global FACES
+    oldfaces, FACES = FACES, altfaces
+    try f()
+    finally
+        FACES = oldfaces
+    end
+end
+
+withfaces(f) = f()
+
+## Face combination and inheritance ##
+
+"""
+    merge(a::Face, b::Face)
+Merge the properties of the text faces `a` and `b`, with those
+of `b` taking priority.
+"""
+function merge(a::Face, b::Face)
+    if isempty(a.inherit)
+        Face(ifelse(isnothing(b.font),          a.font,          b.font),
+             ifelse(isnothing(b.foreground),    a.foreground,    b.foreground),
+             ifelse(isnothing(b.background),    a.background,    b.background),
+             ifelse(isnothing(b.weight),        a.weight,        b.weight),
+             ifelse(isnothing(b.slant),         a.slant,         b.slant),
+             ifelse(isnothing(b.underline),     a.underline,     b.underline),
+             ifelse(isnothing(b.strikethrough), a.strikethrough, b.strikethrough),
+             ifelse(isnothing(b.inverse),       a.inverse,       b.inverse),
+             b.inherit)
+    else
+        merge(merge(a), b)
+    end
+end
+
+"""
+    merge(faces::Face...)
+    merge(faces::Vector{Face})
+
+Merge the properties of `faces` into a single face, with the properties
+of subsequent faces overriding the properties of prior faces.
+"""
+merge(a::Face, b::Face, others::Face...) = merge(merge(a, b), others...)
+merge(faces::Vector{Face}) = foldl(merge, faces, init=Face())
+
+"""
+    merge(face::Face)
+
+Merge any inherited faces of `face` into a self-contained [`Face`](@ref).
+"""
+function merge(face::Face)
```

Comment:
Likewise here, `merge` already has a meaning. If a function for canonicalizing a face is needed, then I think it's better to use another name.

Reply:
With respect to `Face`s, `merge` has in my mind a single meaning: merge all faces, implicit and explicit. The purpose of this single-arity method is to merge the implicit faces within `face` (i.e. inherited values) to a single stand-alone `Face`. I see this as fitting with the spirit of `merge`, but am quite willing to entertain other approaches.

Reply:
In a way, there's almost a single method here: `merge(faces::Face...)` and the fact that we actually have:

```
merge(a::Face)
merge(a::Face, b::Face)
merge(a::Face, b::Face, others::Face...)
``` 

is more an implementation detail than anything else (I'm still thinking about the comment you made about `Vector{Face}`, so let's leave that aside here).

Reply:
It's a bit hard to actually separate those two operations, in order to merge several faces I think you need to canonicalise them.

Reply:
If there's a rule `forall x, merge(x) == x` does this method break it?

Reply:
Hmm, it wouldn't hold `∀x`, or rather any `x` where `x.inherit` is populated.

As mentioned though, the root problem in trying to have a merge-but-don't-canonicalise method is that the two are mixed. The base case (`merge(x)`) is just a special case of the general behavior.

Reply:
About to push a change that I think will have both of us happy.